### PR TITLE
Instrument pypy raw

### PR DESCRIPTION
**Pull request by vext01**
Write out raw PyPy events into results file, e.g.:

Large comment in the code explains the format.

```
 "inst_data": {
  "nbody:PyPy:default-python": {
   "raw_vm_events": [
    [
     [
      "root", 
      null, 
      null, 
      [
       [
        "gc-set-nursery-size", 
        32583494957343, 
        32583495128488, 
        []
       ], 
       [
        "gc-minor", 
        32583495215578, 
        32583495272469, 
        [
         [
          "gc-minor-walkroots", 
          32583495222035, 
          32583495227446, 
          []
         ]
        ]
       ], 
       [
        "gc-set-nursery-size", 
        32583495284669, 
        32583495294252, 
        []
       ], 
```

By all means review, but don't merge yet please. I want to add some tests and try it on a reduced version of the warmup experiment.

### CODE REVIEW

**Comment by vext01** [comment]
Er, github is confused. Some of the old PR comments are here...

**Comment by ltratt** [comment]
Looks good to me. Let me know when you've tested, and I'll merge.

**Comment by vext01** [comment]
Added some tests. VMs still building on bencher2...

**Comment by vext01** [comment]
I've just brought this up to date with master, so I can test the changes to the iterations runners...

**Comment by vext01** [comment]
The iterations runners changes appear to work OK. There were some unrelated crashes to do with Oracle VM updates which I will handle in separate PRs.

OK for me to squash @ltratt?

File: iterations_runners/iterations_runner.py [hunk 1]
```diff
@@ -45,6 +47,11 @@
         bench_func(param)
         stop_time = clock_gettime_monotonic()
 
+        # In instrumentation mode, write a iteration separator to stderr.
```
Comment (ltratt) [inline iterations_runners/iterations_runner.py]: "an iteration separator"

File: krun/results.py [hunk 2]
```diff
@@ -92,6 +100,12 @@ def write_to_file(self):
             f.write(json.dumps(to_write,
                                indent=1, sort_keys=True, encoding='utf-8'))
 
+    def add_instrumentation_data(self, bench_key, inst_dct):
```
Comment (ltratt) [inline krun/results.py]: This method name is a real mouthful... Can we do `add_instr_data` or similar?

**Comment by ltratt** [comment]
There are large numbers of:

```python
    def parse_instrumentation_stderr_file(self, file_handle):
        pass
```

which suggests to me the superclass should have this method as the default?

**Comment by ltratt** [comment]
OK, a few minor comments on the new commits.

**Comment by vext01** [comment]
The `pass` methods are implementing abstract methods. This is a design decision. I am forcing the implementer to think about what needs to be done to support a new VM.

**Comment by ltratt** [comment]
OK, but if they're all `pass`...?

**Comment by vext01** [comment]
Java and PyPy will have implementations of `parse_instrumentation_stderr_file`.

**Comment by ltratt** [comment]
I dunno, I'm not entirely convinced in this case, because if someone doesn't provide a `parse_instr_stderr` (which is a snappier name too) method, then things still work OK? In other words, it'll be pretty obvious to them if they haven't dealt with things.

Maybe a more fundamental question just raised itself to me: is this parsing happening while the iterations runner is happening? Or is it post-processing? Or some combination?

**Comment by vext01** [comment]
If an implementation isn't provided, things will work. Maybe it isn't a huge deal in this case...

Parsing is all post.

**Comment by ltratt** [comment]
OK, so let's make it a `pass` method in the superclass then? [Can we shorten the name at the same time too? Long method names make code harder to read IMHO, as it causes lots of line spills.]

**Comment by vext01** [comment]
Way ahead of you ;)

**Comment by ltratt** [comment]
I expected no less :)

**Comment by vext01** [comment]
Fixed. Caught a bug and some more items that should be `instr_...`. OK to squash?

**Comment by vext01** [comment]
Not sure why GH put my comment in the middle of the changeset :P

**Comment by ltratt** [comment]
Looks good... squash and I'll merge?

**Comment by vext01** [comment]
done

**Comment by ltratt** [comment]
Something just changed? I think there was a second force push.

**Comment by vext01** [comment]
yeah, fixed the test

**Comment by vext01** [comment]
I should have pushed the new commit, so you could see what changed.

There was a test exercising the bogus assertion that I fixed in the last round. I removed this.

**Comment by ltratt** [comment]
OK, mistakes happen, no probs.

**Pull request merged**